cmd/egress-cni-plugin/egressContext.go

```diff
@@ -0,0 +1,393 @@
+	Procsys       procsyswrapper.ProcSys
+	Ipam          hostipamwrapper.HostIpam
+	Link          netlinkwrapper.NetLink
+	Ns            nswrapper.NS
+	NsPath        string
+	ArgsIfName    string
+	Veth          vethwrapper.Veth
+	IpTablesIface iptableswrapper.IPTablesIface
+	IptCreator    func(iptables.Protocol) (iptableswrapper.IPTablesIface, error)
+
+	NetConf   *NetConf
+	Result    *current.Result
+	TmpResult *current.Result
+	Log       logger.Logger
+
+	Mtu int
+	// SnatChain is the chain name for iptables rules
+	SnatChain string
+	// SnatComment is the comment for iptables rules
+	SnatComment string
+}
+
+// EgressContextAddOption has all necessary fields for container egress setup
+type EgressContextAddOption struct {
+	Procsys       procsyswrapper.ProcSys
+	Ipam          hostipamwrapper.HostIpam
+	Link          netlinkwrapper.NetLink
+	Ns            nswrapper.NS
+	NsPath        string
+	ArgsIfName    string
+	Veth          vethwrapper.Veth
+	IpTablesIface iptableswrapper.IPTablesIface
+	IptCreator    func(iptables.Protocol) (iptableswrapper.IPTablesIface, error)
+}
+
+// EgressContextDelOption has all necessary fields for container egress removal
+type EgressContextDelOption struct {
+	Ipam          hostipamwrapper.HostIpam
+	Link          netlinkwrapper.NetLink
+	Ns            nswrapper.NS
+	NsPath        string
+	IpTablesIface iptableswrapper.IPTablesIface
+	IptCreator    func(iptables.Protocol) (iptableswrapper.IPTablesIface, error)
+}
+
+// NewEgressAddContext create a context for container egress traffic
+func NewEgressAddContext(option EgressContextAddOption) (*EgressContext, error) {
+	if option.Procsys == nil {
+		return nil, fmt.Errorf("prosys is nil in EgressContextAddOption")
+	}
+	if option.Ipam == nil {
+		return nil, fmt.Errorf("ipam is nil in EgressContextAddOption")
+	}
+	if option.Link == nil {
+		return nil, fmt.Errorf("link is nil in EgressContextAddOption")
+	}
+	if option.Ns == nil {
+		return nil, fmt.Errorf("ns is nil in EgressContextAddOption")
+	}
+	if option.NsPath == "" {
+		return nil, fmt.Errorf("nsPath is empty in EgressContextAddOption")
+	}
+	if option.ArgsIfName == "" {
+		return nil, fmt.Errorf("argsIfName is empty in EgressContextAddOption")
+	}
+	if option.Veth == nil {
+		return nil, fmt.Errorf("veth is nil in EgressContextAddOption")
+	}
+	if option.IpTablesIface == nil && option.IptCreator == nil {
+		return nil, fmt.Errorf("both ipTablesIface and iptCreator are nil in EgressContextAddOption")
+	}
+	return &EgressContext{
+		Procsys:       option.Procsys,
+		Ipam:          option.Ipam,
+		Link:          option.Link,
+		Ns:            option.Ns,
+		NsPath:        option.NsPath,
+		ArgsIfName:    option.ArgsIfName,
+		Veth:          option.Veth,
+		IpTablesIface: option.IpTablesIface,
+		IptCreator:    option.IptCreator,
+	}, nil
+}
+
+// NewEgressDelContext create a context for container egress traffic
+func NewEgressDelContext(option EgressContextDelOption) (*EgressContext, error) {
```

Comment:
Why not just have one? We can just pass in `nil` for those fields we don't plan to use for DEL for example...

Reply:
Also, I don't think there's any need to return an error from this function. Not passing an arg should not be an error

Reply:
Add and Del need different fields and it is good to have their own and do their own check.

Some fields are required and not passing an arg will not work.

Reply:
If we only use one new constructor for both Add/Del and assign nil to not passed fields, the refactor was not better than original code that just do egressContext{ some assignments}

Reply:
Put an update per discussion with @jdn5126
To summary: 
1. We want to have a new constructor — not to instantiate any structs in main.go
2. We don’t want new constructor to have err return
3. If we want to achieve 2, we have to refactor unit testing not to use new constructor, and that’s OK.